### ISSUE TITLE
Automatic Cache Invalidation

### ISSUE THREAD

**Issue by flybayer** [blitz-js/blitz#586]
### What do you want and why?

Currently you need to manually call `refetch()` or [`mutate()`](#582) on a `useQuery` result after doing a mutation.

We should try to come up with a way to automatically invalidate queries so that most users don't have to think about it all.

I haven't spent a lot of time thinking about this, but here's a couple options.

### Based on file structure

We could automatically invalidate queries at the same folder level of the mutation.

So with this file structure,
```
app/products/queries/getProduct.ts
app/products/queries/getProducts.ts
app/products/mutations/createProduct.ts
app/products/mutations/updateProduct.ts
app/products/mutations/deleteProduct.ts
```

Calling `createProduct`, `updateProduct`, or `deleteProduct` would invalidate the cache for `getProduct` and `getProducts`.

But the problem with this is that `app/queries/getDisplayProducts.ts` would not be invalidated even thought it is also affected. But maybe this is still worth it if it solves the 80% use case.

### Via Some Type of Introspection Magic

Maybe it's possible for us to compute the entire data graph at build time using a combination of the Prisma DMMF and Typescript? If possible, this would be amazing because we could accurately invalidate all queries anywhere in the app.

I think this would only work with Prisma, so we should still have some other fallback method for Blitz apps without Prisma.

### On Route Change

Something else that's a possibility is [busting the cache on route change](#583), but that still requires you to think about it.

**I'm super keen to hear anyone's ideas on this!**

**Comment by ryardley** [comment]
The problem I see with file based metadata is that it cannot reference related models. Is decorators on queries and mutations out of the question?

**Comment by flybayer** [comment]
@ryardley I think anything is fair game and worth exploring!

**Comment by flybayer** [comment]
@ryardley I think we should go ahead and add this part of automatic query invalidation. It's trivial to implement.

> We could automatically invalidate queries at the same folder level of the mutation.
> 
> So with this file structure,
> 
> ```
> app/products/queries/getProduct.ts
> app/products/queries/getProducts.ts
> app/products/mutations/createProduct.ts
> app/products/mutations/updateProduct.ts
> app/products/mutations/deleteProduct.ts
> ```
> 
> Calling `createProduct`, `updateProduct`, or `deleteProduct` would invalidate the cache for `getProduct` and `getProducts`.

**Comment by ryardley** [comment]
One thing that concerns me with this is that is suggests that people should break their apps into many small contexts which might be an architectural mistake. Contexts should be DDD contexts which probably hold a few different models and not all of them necessarily should be invalidated. We might need to group mutations into resources somehow (maybe folders) and then invalidate based on that? Or possibly provide decorators or provide some other metadata for connecting m/q's together?

**Comment by flybayer** [comment]
@ryardley yeah makes sense. My thought is that over-eager cache busting is better than none at all. We can always optimize this later.

**Comment by alexturpin** [comment]
👋 I'm far from an expert but I've been reading through the docs and wondering how to do this myself. What about passing in a query to the `useMotation` hook and that query gets invalidated once a mutation has succeeded?

**Comment by flybayer** [comment]
@alexturpin yeah, that's a good idea! Not exactly automatic, but I that would be an improvement.

**Comment by NICHTJ3** [comment]
What about a list of queries kind of like dependencies in a useEffect hook?

**Comment by NICHTJ3** [comment]
Optionally declare the list as a named export in the query or mutation file to make it somewhat more automatic. (or even a list of linked models to invalidate related queries from that way we aren't over eager?)

I think the explicitness of declaring them as dependencies of the hook directly lets other developers know what queries/models are dependent without having to jump to a different context, but I also imagine writing it out a thousand times would be time consuming/frustrating

**Comment by 8balloon** [comment]
@flybayer As a preface, I spent a lot of time on this problem while prototyping my own "zero-api" solution before I found Blitz :) so hopefully this is useful.

### Based on file structure convention

This is an essential component of the "file dependencies + conventions" option detailed below

### Based on Prisma introspection magic

This could be useful for optimizing a more general solution

### On route change

Thanks for doing this! :)

### Dependecy lists

There's no obvious or testable way to determine whether a dependency list is complete, which makes these approaches difficult to maintain IME

## New ideas

### Based on file dependencies + conventions (Mobx-like)

All queries at the same folder level of the mutation are invalidated automatically. Additionally, any other folders containing a query that references the mutation's folder (directly or indirectly) also have all of their queries invalidated.

The idea behind this is that the dependency graph would be represented by the _actual project file dependencies_ rather than the Prisma schema.

Problem A: using nested relational access via Prisma would break automatic query invalidation, since it does not rely on file structure.

Solution to Problem A: extend `enhancePrisma` to prevent nested relational access. To access related entities in a query while preserving proper cache invalidation, users must compose queries from other queries (which basically amounts to a repository or a service pattern).

Upsides: a single, enforced community Repository pattern with clear separation of concerns and high maintainability (which could help set it apart from Rails/ActiveRecord)

Downsides: 1) The Blitz community would have to be sold on "composition good, nested access bad" 2) cache invalidation could still be broken if the repository pattern was circumvented

### "shouldCacheInvalidate" (like "shouldComponentUpdate")

Whenever any mutation occurs, invalidate all queries currently subscribed to by rendered views. For performance optimizations, allow queries to define a `shouldCacheInvalidate` function, like so:

```js
import createChoice from "app/choices/mutations/createChoice"

const getQuestion = resolver.pipe(resolver.zod(GetQuestion), resolver.authorize(), ({ id }) => {
  return db.question.findFirst({ where: { id } })
})

getQuestion.shouldCacheInvalidate = (mutation, input, ctx) => {
  return mutation !== createChoice
}

export default getQuestion
```

This would be fully automatic and reliable (even for those leveraging nested queries), and relatively easy to optimize.

An implementation of this might look like so:
  - when the client notifies the server that it's performing a mutation, it also sends it a list of `[query, queryInput, hash(cachedQueryResult)]` objects corresponding to all query/queryInput combinations it's subscribed to. The server runs the mutation, reruns all queries with corresponding query inputs, and sends back a list of `[query, hash(queryInput), queryResult]` objects, excluding results from the list where `hash(queryResult) === hash(cachedQueryResult)`. The FE cache is then updated with the new `queryResult`s.

Thanks everyone, especially @flybayer for the awesome framework. I'm excited to see how things continue to evolve! Also I'm so happy that Blitz is so eminently usable, even without automatic cache invalidation :pray:

**Comment by flybayer** [comment]
@8balloon thank you, I love it!

I've caught wind of some possible caching changes in react-query that could be very useful here, so I think we should sit on this until we get more details on any react-query changes.

**Comment by jan-wilhelm** [comment]
> I've caught wind of some possible caching changes in react-query that could be very useful here
Any links that you could point us to? :) sounds awesome

**Comment by flybayer** [comment]
Nope

**Comment by oceandrama** [comment]
Highly anticipated feature

**Comment by MadcowD** [comment]
Is this dead?

**Comment by beerose** [comment]
> Is this dead?

It's not! We just haven't had the time to work on it yet.
